lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs

```diff
@@ -0,0 +1,192 @@
+            timer = new Timer(timerIntervalSec * 1000);
+            timer.Elapsed += OnTimer;
+            timer.AutoReset = true;
+            timer.Start();
+        }
+
+        public IDisposable Subscribe(IObserver<IJobCancelled> observer)
+        {
+            if (observer == null)
+            {
+                throw new ArgumentNullException("observer");
+            }
+
+            lock (_observers)
+            {
+                if (!_observers.Contains(observer))
+                {
+                    _observers.Add(observer);
+                }
+            }
+
+            return new AnonymousDisposable(() => Unsubscribe(observer));
+        }
+
+        public void Dispose()
+        {
+            if (this.timer != null)
+            {
+                lock (lockDispose)
+                {
+                    if (timer != null)
+                    {
+                        timer.Stop();
+                        timer.Dispose();
+                        timer = null;
+                    }
+                }
+            }
+        }
+
+        public void OnError(Exception error)
+        {
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        private void Unsubscribe(IObserver<IJobCancelled> observer)
+        {
+            lock (_observers)
+            {
+                _observers.Remove(observer);
+            }
+        }
+
+        private void OnTimer(object source, ElapsedEventArgs e)
+        {
+            if (!_observers.Any())
+            {
+                Logger.Log(Level.Warning, "There are no observers for cancellation event: skipping cancellation detection");
+                return;
+            }
+
+            string cancellationMessage = null;
+            if (IsJobCancelled(out cancellationMessage))
+            {
+                Logger.Log(Level.Info, "Detected Job cancellation ({0}): sending JobCancelled event to observers: {1}", cancellationMessage, _observers);
+                var cancelEvent = new JobCancelled(DateTime.Now, cancellationMessage);
+
+                lock (_observers)
+                {
+                    _observers.ForEach(o => o.OnNext(cancelEvent));
+                }
+            }
+        }
+
+        private bool IsJobCancelled(out string cancellationMessage)
+        {
+            var isCancelled = false;
+            cancellationMessage = null;
+
+            try
+            {
+                isCancelled = CancellationDetector != null && CancellationDetector.IsJobCancelled(out cancellationMessage);
+            }
+            catch (Exception ex)
+            {
+                Logger.Log(Level.Error, "IsCancelled check failed. Exception:{0}", ex);
+                isCancelled = false;
+            }
+
+            return isCancelled;
+        }
+
+        [NamedParameter("Sleep interval for Job lifecycle manager", "lifecyclesleepintervalSec", "15")]
+        public sealed class SleepIntervalParameter : Name<int>
```

Comment:
done

---
In reply to: [84164424](https://github.com/apache/reef/pull/1161#discussion_r84164424) [](ancestors = 84164424)